RFC-0007-plan-constraints-grammar.md

```diff
@@ -0,0 +1,77 @@
+# **RFC007 for Presto**
+
+## Grammar for Plan Constraints
+
+Proposers
+
+* @aaneja
+* @ClarenceThreepwood
+
+## Related Issues
+
+* [Overview doc](https://prestodb.io/wp-content/uploads/Search-Space-Improvements-Plan-Constraints.pdf) on Plan Constraints as a tool to control search space
+* PrestoDB blog on - [Elevating Presto Query Optimization](https://prestodb.io/blog/2024/03/21/elevating-presto-query-optimization/)
+
+## Summary
+
+This document proposes a grammar for specifying plan constraints
+
+## Background
+
+Plan constraints can be used to lock down critical aspects of an execution plan, such as access method, join method, and join order. 
+
+
+## Proposed Implementation
+
+We propose a grammar for specifying independent plan constraints, which take the form of a SQL comment block that would build an object graph of the constraints. Multiple constraints can be specified in a single place. The grammar is open for extension as we develop more mechanisms to lock down plans.
+
+In this first cut, users can build constraints to control 
+- Join orders and distributions for INNER JOIN's
+- Cardinality (row counts) for base relations and join sub-plans
+
+### Grammar
+
+```
+planConstraintString : /*! planConstraint [, ...] */
+
+planConstraint : joinConstraint
+| cardinalityConstraint
+
+joinConstraint : joinType (joinNode) [distributionType]
+
+cardinalityConstraint : CARD (joinNode cardinality)
+
+distributionType : [P]
+   | [R]
+
+joinType : JOIN	(defaults to inner join)
+| IJ
+| LOJ
+| ROJ
+
+cardinality : integer constant (positive)
+
+joinNode : (relationName relationName [, ...])
+
+| joinConstraint
+
+| (relationName joinNode)
+
+| (joinNode relationName)
+
+| (joinNode joinNode [, ...])
+```
+
+
+### Examples of constraints
+
+1. Inner Join constraints - 
+    1. `join (a (b c))` - Join the relations a,b,c as a right deep tree (denoted by the brackets). Use regular rules for determining join distribution
+    2. `join (((a c) [R] b) [P])` - In addition to the join order, use a  REPLICATED `[R]`  join for sub-plan `(a c)`  and PARTITIONED `[P]`  for `(a c) b`
+    3. If an inner join condition does not exist between nodes a CrossJoin is automatically inferred
+2. Cardinality constraints -
+    1. `card (c 10)` - Set the output row count estimate of `c` to `10` 
+    2. `card ((c o) 10)` - When considering a join node of shape `(c o)` set the output row count estimate to `10`
+
+### Other points of note
+- Relation names loosely resolve to WITH query aliases (CTE definitions) and table names.A detailed description of the name resolution is out of scope of this RFC (this will be covered in the implementation PR description)
```

Comment:
It might be worth fleshing out in more detail exactly how we will deal with name resolution and name scoping for constraints. I'm mainly concerned about name clashes between CTEs/subqueries/aliases (I'm not sure of the exact terminology).

Consider the following simple example:

```sql
WITH t as (
    SELECT * FROM
    (VALUES 1, 2) as t(a)
    LIMIT 1)
SELECT * FROM t;
```

How would you write a constraint which would be able to unambiguously reference the `t` from `WITH t` and the `t` from `VALUES ... t(a)`? This is important because each table `t` has a different cardinality `VALUES t` has a cardinality of 2, while `WITH t` has cardinality `1` due to the `LIMIT`. I would imagine we don't want to limit our ability to define constraints when names clash.

Reply:
>I would imagine we don't want to limit our ability to define constraints when names clash.
In the first implementation, we will require users to define unambiguous names for cases like this, i.e they would have to rewrite their SQL.
However, this can improved upon. We can build the name matcher to allow us to specify nested names like `t.t`.
Since the name matching is implementation dependent, and does not require a grammar rule, I chose to skip describing it here. I will be covering it in the implementation PR

Reply:
Is there a case to refer to something internal to a with clause in the outer context ? Generally, I would imagine constaints to be placed within the with definition itself.

Using an outer context is more complicated since there could be multiple uses of the CTE and so we would have to think about disambiguating the uses of the CTE as well.